### ISSUE TITLE
Function pointers are difficult to parse in vk.xml

### ISSUE THREAD

**Issue by Kangz** [KhronosGroup/Vulkan-Docs#83]
Vk.xml is a nice but there are a couple spots where it is hard to parse, the worst being function pointer types definition. Here is an example:
```xml
        <type category="funcpointer">typedef void (VKAPI_PTR *<name>PFN_vkInternalAllocationNotification</name>)(
    <type>void</type>*                                       pUserData,
    <type>size_t</type>                                      size,
    <type>VkInternalAllocationType</type>                    allocationType,
    <type>VkSystemAllocationScope</type>                     allocationScope);</type>
```
Getting the return type of the function and splitting the parameters is not easy nor is splitting parameters. (this becomes even worse when a `const` gets before the `<type>` tag)

I suggest using the same structure as `<command>` tags that clearly separate the prototype (and return type) as well as the parameters:

```xml
        <command queues="graphics" renderpass="inside" cmdbufferlevel="primary,secondary">
            <proto><type>void</type> <name>vkCmdDrawIndirect</name></proto>
            <param externsync="true"><type>VkCommandBuffer</type> <name>commandBuffer</name></param>
            <param><type>VkBuffer</type> <name>buffer</name></param>
            <param><type>VkDeviceSize</type> <name>offset</name></param>
            <param><type>uint32_t</type> <name>drawCount</name></param>
            <param><type>uint32_t</type> <name>stride</name></param>
            <validity>
                [skipped]
            </validity>
        </command>
```

**Comment by oddhack** [comment]
Thanks. This is on my todo list, but at a relatively low priority given all the post-release stuff we're handling - if someone wanted to propose tweaks to the XML syntax and scripts that accomplish this, it would be cool.

**Comment by SilverPhoenix99** [comment]
Hi.

I hope you don't mind me linking to https://github.com/KhronosGroup/Vulkan-Docs/issues/19#issuecomment-186182788, since it's essentially the same issue.

**Comment by NicolBolas** [comment]
The issue in #19 is that it's hard to work with the data in the `command` element. The issue here is that the function pointer `type` isn't as easy to work with as `command`. They're really talking about two different problems.

**Comment by SilverPhoenix99** [comment]
@NicolBolas in #19 I suggested adding attributes to `command` and elements to `funcpointer` types. This issue is also about adding elements to `funcpointer` types in the same way that was suggested in #19, which is (non) basically transform it into `command` types. Of course, this implies lots of changes to the spec, the RNG schema, and the header generator.

@Kangz, I don't know what is your target language, but what I've done with `funcpointer` types is to read the following text element and delete the `const` keyword explicitly after the comma. Not pretty, but it works.

**Comment by timofonic** [comment]
@Kangz @oddhack @NicolBolas @SilverPhoenix99 Any news?

**Comment by NogginBops** [comment]
@oddhack is this something that is possible to do or has the time for this change passed?
What scripts would need to change for a change like this to be merged?

**Comment by oddhack** [comment]
The problem with making schema changes that are not purely additive at this point is that there are numerous downstream consumers, particularly the LunarG SDK which is critical to the ecosystem. It's not just our scripts, it's all the components and language bindings, several of which don't every use out Python scripts and thus have to re-write their own parsers in C++ or Rust or whatever.

The last time we did this, for the 'depends' schema change, it took approaching a year to get everyone lined up to support it. So it's a very hard problem and more trouble than we're usually willing to take on. The only potentially viable approach I've ever thought of is to  have a backwards-compatibility version of the XML that is automatically transformed from the updated schema, something that might not even be possible without loss of information and practically speaking, would never be retired.

**Comment by NogginBops** [comment]
That's unfortunate. And I don't really see a solution where we can expose this extra information in an additive way without breaking people (Can't do nested thing, or alternative category because downstream users will not be able to parse those).

To be clear what I want is just `<param>` and `<name>` tags to be added:
```xml
<type category="funcpointer">typedef <type>void</type> (VKAPI_PTR *<name>PFN_vkInternalAllocationNotification</name>)(
    <param><type>void</type>*                                       <name>pUserData</name></param>,
    <param><type>size_t</type>                                      <name>size</name></param>,
    <param><type>VkInternalAllocationType</type>                    <name>allocationType</name></param>,
    <param><type>VkSystemAllocationScope</type>                     <name>allocationScope</name>)</param>;</type>
```
but I'm guessing the breaking chance for downstream users is high enough that it's not worth it?

Making a backwards compatibility version of the XML seems like a lot of work, but I guess it could be worth it to be able to actually be able to make improvements to the XML specification.

I guess I'll have to work around this, not impossible but slightly annoying. Should this issue be closed then if this isn't going to be solved?
